build/generate-catalog-template.sh

```diff
@@ -0,0 +1,70 @@
+#! /bin/bash
+
+set -e
+
+if [[ $(basename "${PWD}") != "gatekeeper-operator-fbc" ]]; then
+  echo "error: Script must be run from the base of the repository."
+  exit 1
+fi
+
+echo "Using drop version OCP-Gatekeeper map:"
+jq '.' drop-versions.json
+
+ocp_versions=$(jq -r 'keys[]' drop-versions.json)
+
+shouldPrune() {
+  oldest_version="$(jq -r ".[\"${1}\"]" drop-versions.json).99"
+
+  [[ "$(printf "%s\n%s\n" "${2}" "${oldest_version}" | sort --version-sort | tail -1)" == "${oldest_version}" ]]
+
+  return $?
+}
+
+for version in ${ocp_versions}; do
+  cp catalog-template.yaml "catalog-template-${version//./-}.yaml"
+done
+
+# Prune old X.Y channels
+echo "# Pruning channels:"
+for channel in $(yq '.entries[] | select(.schema == "olm.channel").name' catalog-template.yaml); do
+  echo "  Found channel: ${channel}"
+  for ocp_version in ${ocp_versions}; do
+    if shouldPrune "${ocp_version}" "${channel}"; then
+      echo "  - Pruning channel from OCP ${ocp_version}: ${channel} ..."
+      yq '.entries[] |= select(.schema == "olm.channel") |= del(select(.name == "'"${channel}"'"))' -i "catalog-template-${ocp_version//./-}.yaml"
+
+      continue
+    fi
+
+    # Prune old bundles from channels
+    for entry in $(yq '.entries[] | select(.schema == "olm.channel") | select(.name == "'"${channel}"'").entries[].name' catalog-template.yaml); do
+      version=${entry#*\.v}
+      if shouldPrune "${ocp_version}" "${version}"; then
+        echo "  - Pruning entry from OCP ${ocp_version}: ${entry}"
+        yq '.entries[] |= select(.schema == "olm.channel") |= select(.name == "'"${channel}"'").entries[] |= del(select(.name == "'"${entry}"'"))' -i "catalog-template-${ocp_version//./-}.yaml"
+      fi
+    done
+  done
+done
+echo
+
+# Prune old bundles
+echo "# Pruning bundles:"
+for bundle_image in $(yq '.entries[] | select(.schema == "olm.bundle").image' catalog-template.yaml); do
+  bundle_version=$(skopeo inspect --override-os=linux --override-arch=amd64 "docker://${bundle_image}" | jq -r ".Labels.version")
+  echo "  Found version: ${bundle_version}"
+  pruned=0
+  for ocp_version in ${ocp_versions}; do
+    if shouldPrune "${ocp_version}" "${bundle_version#v}"; then
+      echo "  - Pruning bundle ${bundle_version} from OCP ${ocp_version} ..."
+      echo "    (image ref: ${bundle_image})"
+      yq '.entries[] |= select(.schema == "olm.bundle") |= del(select(.image == "'"${bundle_image}"'"))' -i "catalog-template-${ocp_version//./-}.yaml"
+    else
+      ((pruned+=1))
+    fi
+  done
+  if ((pruned == $(jq 'keys | length' drop-versions.json))); then
+    echo "Nothing pruned--exiting."
+    break
+  fi
```

Comment:
I'm not clear on this, couldn't the script need to prune more than this number? Like, if I'm trying to drop all bundles older than (just for example) 3.15, I might only have that one entry in `drop-versions.json`, but I want to remove `3.10`, `3.11`, `3.12`, etc, right?

Overall, this doesn't need to block this PR. The script can be adjusted later. Everything it does is local, and then it gets committed to git, right? So we can always easily review exactly what it does.

Reply:
Right--the bundles are listed in order though, so once it stops pruning, the rest wouldn't be pruned. Though it could keep running for completeness. The `skopeo` commands take a bit, so I was short-circuiting it for speed (though it'll be less painful for the next person, so I could take it out).